pkg/backend/display/progress.go

```diff
@@ -661,99 +668,203 @@ func (display *ProgressDisplay) processEndSteps() {
 		}
 	}
 
-	// Now refresh everything.  this ensures that we go back and remove things like the diagnostic
+	// Now refresh everything.  This ensures that we go back and remove things like the diagnostic
 	// messages from a status message (since we're going to print them all) below.  Note, this will
-	// only do something in a terminal.  This i what we want, because if we're not in a terminal we
+	// only do something in a terminal.  This is what we want, because if we're not in a terminal we
 	// don't really want to reprint any finished items we've already printed.
 	display.refreshAllRowsIfInTerminal()
 
-	// Print all diagnostics we've seen.
+	// Render several "sections" of output based on available data as applicable.
+	display.writeBlankLine()
+	display.printDiagnostics()
+	display.printPolicyViolations()
+	display.printOutputs()
+	display.printSummary()
+}
 
-	wroteDiagnosticHeader := false
+var wroteDiagnosticHeader bool
```

Comment:
I access this later in printSummary(), is there a better way to achieve this?

Reply:
Gotcha. Well it's still "-10,000 points" for having a global variable like this, that is modified as a side effect of calling methods on a given type, etc. But here's some ways to approach refactoring the code so that this is a little less error prone:

But if we need to rely on this value in `printSummary()`, then let's think about _why_. Does `printSummary()` need to know if a diagnostic header has been written or not for some reason? Then rather than rely on some ambient state, we should probably just pass in the required data as a parameter.

e.g.

```go
	wroteDiagnosticHeader := display.printDiagnostics()
	display.printDiagnostics()
 	display.printPolicyViolations()
 	display.printOutputs()
 	display.printSummary(wroteDiagnosticHeader)
```

... but it seems oddly specific that "printing the summary" depends on whether or not we emitted a diagnostic header. Perhaps that parameter name should be renamed to something more meaningful. Perhaps:

```
func (...) printSummary(insertExtraBlankLine bool) {
    ...
}
```

... and if it is something super-straight forward, like whether or not we need to emit an extra blank line or something, then we can probably take care of that _before_ we call `printSummary()`. So that `printSummary()` is no longer "coupled" or dependent on that value. e.g.

```go
wroteDiagnostic := display.printDiagnostics()
...

// If we wrote diagnostics out, we <need to do X> <because of Y>.
if wroteDiagnostics {
    doY()
}
display.printSummary()
```

Reply:
Great summary by @chrsmith !  This would definitely be a preferred way to do this versus having a global variable.

Another form that could be used (though people may love/hate would be):

```go
	var wroteDiagnosticHeader = false
        display.printDiagnostics(&wroteDiagnosticHeader)
	display.printDiagnostics(&wroteDiagnosticHeader)
 	display.printPolicyViolations(&wroteDiagnosticHeader)
 	display.printOutputs(&wroteDiagnosticHeader)
 	display.printSummary(&wroteDiagnosticHeader)
```

The variable now acts as an in/out flag that allows the value to be passed in, but also allows the callee to update it with their updated value.  

These forms help keep the value scoped only to the level where it is needed (As opposed to the global variable) and make it clear who is supposed to use/update it.